File: Go/Problems/4_Anagram_Groups/main_test.go
package main

import (
	"bytes"
	"io"
	testutils "needcode-150/Utilities/testUtils"
	"os"
	"testing"
)

func TestMain(t *testing.T) {
	// Save the original stdout
	originalStdout := os.Stdout
	r, w, _ := os.Pipe()
	os.Stdout = w

	// Call the main function
	main()

	// Capture the output
	w.Close()
	var buf bytes.Buffer
	io.Copy(&buf, r)
	os.Stdout = originalStdout

	// Check the output
	expected := "Hello, Anagram_Groups!\n"
	if buf.String() != expected {
		t.Errorf("expected %q but got %q", expected, buf.String())
	}
}

// eg. 1
// Input: strs = ["act","pots","tops","cat","stop","hat"]

// Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]

// eg. 2
// Input: strs = ["x"]
// Output: [["x"]]

// eg. 3
// Input: strs = [""]

// Output: [[""]]

func TestGroupAnagram(t *testing.T) {
	testCases := []struct {
		strs     []string
		expected [][]string
	}{
		{[]string{"act", "pots", "tops", "cat", "stop", "hat"}, [][]string{{"hat"}, {"pots", "tops", "stop"}, {"act", "cat"}}},
		{[]string{"x"}, [][]string{{"x"}}},
		{[]string{""}, [][]string{{""}}},
	}

	for _, tc := range testCases {
		output := GroupAnagram(tc.strs)
		if !testutils.CompareStringSlices(output, tc.expected) {
			t.Errorf("expected %v but got %v", tc.expected, output)
		}
	}
}

File: JS/Problems/4_Anagram_Groups/main.test.js
const Solution = require('./main'); // Adjust the path if necessary

test('main function output', () => {
    const solution = new Solution();
    const output = solution.helloWorld()
    console.log(output)
    const expected = "Hello, World!";
    expect(output).toBe(expected);
});

// eg. 1
// Input: strs = ["act","pots","tops","cat","stop","hat"]

// Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]
test('Eg. 1', () => {
    const solution = new Solution();
    const output = solution.groupAnagrams(["act","pots","tops","cat","stop","hat"])
    console.log(output)
    const expected = [["act","cat"],["pots","tops","stop"],["hat"]];
    expect(output).toStrictEqual(expected);
});

// eg. 2
// Input: strs = ["x"]
// Output: [["x"]]

test('Eg. 2', () => {
    const solution = new Solution();
    const output = solution.groupAnagrams(["x"])
    console.log(output)
    const expected = [["x"]];
    expect(output).toStrictEqual(expected);
})

// eg. 3
// Input: strs = [""]

// Output: [[""]]
test('Eg. 3', () => {
    const solution = new Solution();
    const output = solution.groupAnagrams([""])
    console.log(output)
    const expected = [[""]];
    expect(output).toStrictEqual(expected);
})
